---
title: "microbiome"
author: "Tsunghan Hsieh"
date: "2023-01-14"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Workflow for Microbiome Data Analysis

This is a tutorial document for microbiome data analysis. Credit: <https://bioconductor.org/help/course-materials/2017/BioC2017/Day1/Workshops/Microbiome/MicrobiomeWorkflowII.html#construct_phylogenetic_tree>.

## Packages

```{r packages}
library("knitr")
library("BiocStyle")
.cran_packages <- c("ggplot2", "gridExtra")
.bioc_packages <- c("dada2", "phyloseq", "DECIPHER", "phangorn")
.inst <- .cran_packages %in% installed.packages()
if(any(!.inst)) {
   install.packages(.cran_packages[!.inst])
}
.inst <- .bioc_packages %in% installed.packages()
if(any(!.inst)) {
  if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
  BiocManager::install(.bioc_packages[!.inst], ask = F)
}
# Load packages into session, and print package version
sapply(c(.cran_packages, .bioc_packages), require, character.only = TRUE)
```
## Input data
```{r input_data}
set.seed(1234) # make sure the random sampling data could be replicated in each time
miseq_path <- "/Users/hsieh/Microbiome/Data/MiSeq_SOP"
list.files(miseq_path)
```

## Filter and Trim
### Check data
```{r check_data}
# Sort ensures forward/reverse reads are in same order
fnFs <- sort(list.files(miseq_path, pattern="_R1_001.fastq"))
fnRs <- sort(list.files(miseq_path, pattern="_R2_001.fastq"))

# Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq
sampleNames <- sapply(strsplit(fnFs, "_"), `[`, 1)

# Specify the full path to the fnFs and fnRs
fnFs <- file.path(miseq_path, fnFs)
fnRs <- file.path(miseq_path, fnRs)
fnFs[1:3]
fnRs[1:3]
```

### plot the quality of forward and reverse data
```{r plot_quality, echo=FALSE}
plotQualityProfile(fnFs[1:2]) # after 240, the quality drops
```
```{r plot_quality, echo=FALSE}
plotQualityProfile(fnRs[1:2]) # after 160, the quality drops
```

### Filtering
```{r filtering}
# Generate filtered file names
filt_path <- file.path(miseq_path, "filtered") # Place filtered files in filtered/ subdirectory
if(!file_test("-d", filt_path)) dir.create(filt_path)
filtFs <- file.path(filt_path, paste0(sampleNames, "_F_filt.fastq.gz"))
filtRs <- file.path(filt_path, paste0(sampleNames, "_R_filt.fastq.gz"))

out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(240,160),
              maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE,
              compress=TRUE, multithread=TRUE) # On Windows set multithread=FALSE
head(out)
```

## Infer sequence variants
The basic idea of 16S rRNA sequence analysis is to compare the differences of 16S rRNA sequence between each sample and assign a taxonomy name to the sample. However, errors occurred during sequencing process could wrongly give the sample different taxonomy name. Therefore, we should first calculate the error rate among the whole experiments and set the threshold for excluding sequences with too many errors.

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
